README
                                W32PTH
                  An emulation of PTH for MS Windows
                 ------------------------------------

This is an emulation of GNU PTH for MS Windows (W32).  It works by
blocking all threads when outside of one of the pth functions and
using standard W32 threads and suncronization mechanisms to emulate
the behaviour of PTH.

It is currently limited to what GnuPG 2.0 requires.


Missing stuff:

  PTH_MODE_STATIC is known but ignored.
  PTH_MODE_REUSE
  PTH_MODE_CHAIN

  rwlock - currently just a mutex

  as well as many more things.



NEWS
Noteworthy changes in version 2.0.5 (2013-04-23)
------------------------------------------------

 * Fix regression in W32 timer handling introduced by 2.0.3.

 * Switch to i686-w64-mingw32 as default toolchain.


Noteworthy changes in version 2.0.4 (2010-08-02)
------------------------------------------------

 * Bug fix: Export pth_enter and pth_leave.


Noteworthy changes in version 2.0.3 (2010-07-30)
------------------------------------------------

 * Support WindowsCE.

 * New functions pth_enter and pth_leave.


Noteworthy changes in version 2.0.2 (2008-10-17)
------------------------------------------------

 * Support pipes created with non-standard extensions pth_pipe and
   closed with pth_close.

 * Allow destruction of mutexes with non-standard extension
   pth_mutex_destroy.

 * Support pth_usleep.

 * Bug fixes.

 * New non-standard fucntion pth_thread_id.

 * Add PTH_HANDLE_EVENT.


Noteworthy changes in version 2.0.1 (2007-08-16)
------------------------------------------------

 * Fixes.  Better support for select.


Noteworthy changes in version 2.0.0 (2007-07-05)
------------------------------------------------

 * First release.  We need to use this unusually high version number
   to indicate the version of Pth we are emulating.


 Copyright 2007, 2008 g10 Code GmbH

 This file is free software; as a special exception the author gives
 unlimited permission to copy and/or distribute it, with or without
 modifications, as long as this notice is preserved.

 This file is distributed in the hope that it will be useful, but
 WITHOUT ANY WARRANTY, to the extent permitted by law; without even the
 implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


debug.h
/* debug.h - interface to debugging functions
   Copyright (C) 2002, 2004, 2005, 2007, 2008 g10 Code GmbH
 
   This file is part of PTH.

   PTH is free software; you can redistribute it and/or modify it
   under the terms of the GNU Lesser General Public License as
   published by the Free Software Foundation; either version 2.1 of
   the License, or (at your option) any later version.
   
   PTH is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.
   
   You should have received a copy of the GNU Lesser General Public
   License along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
   02111-1307, USA.  */

#ifndef DEBUG_H
#define DEBUG_H

#include <string.h>


/* Keeps the current debug level. Define marcos to test them. */
extern int debug_level;
#ifdef HAVE_W32CE_SYSTEM
extern HANDLE dbghd;
#else
extern FILE *dbgfp;
#endif
#define DBG_ERROR  (debug_level >= 1)
#define DBG_INFO   (debug_level >= 2)
#define DBG_CALLS  (debug_level >= 3)

#define DEBUG_ERROR	1
#define DEBUG_INFO	2
#define DEBUG_CALLS	3


/* Indirect stringification, requires __STDC__ to work.  */
#define STRINGIFY(v) #v
#define XSTRINGIFY(v) STRINGIFY(v)

/* Log the formatted string FORMAT at debug level LEVEL or higher.  */
void _pth_debug (int level, const char *format, ...);


/* Trace support.  */

/* FIXME: For now.  */
#define _pth_debug_trace() 1

#define _TRACE(lvl, name, tag)					\
  int _pth_trace_level = lvl;					\
  const char *const _pth_trace_func = name;			\
  const char *const _pth_trace_tagname = STRINGIFY (tag);	\
  void *_pth_trace_tag = (void *) tag

#define TRACE_BEG(lvl, name, tag)			   \
  _TRACE (lvl, name, tag);				   \
  _pth_debug (_pth_trace_level, "%s (%s=0x%x): enter\n", \
		_pth_trace_func, _pth_trace_tagname,   \
		_pth_trace_tag), 0
#define TRACE_BEG0(lvl, name, tag, fmt)					\
  _TRACE (lvl, name, tag);						\
  _pth_debug (_pth_trace_level, "%s (%s=0x%x): enter: " fmt "\n",	\
		_pth_trace_func, _pth_trace_tagname,		\
		_pth_trace_tag), 0
#define TRACE_BEG1(lvl, name, tag, fmt, arg1)				\
  _TRACE (lvl, name, tag);						\
  _pth_debug (_pth_trace_level, "%s (%s=0x%x): enter: " fmt "\n",	\
		_pth_trace_func, _pth_trace_tagname,		\
		_pth_trace_tag, arg1), 0
#define TRACE_BEG2(lvl, name, tag, fmt, arg1, arg2)		    \
  _TRACE (lvl, name, tag);					    \
  _pth_debug (_pth_trace_level, "%s (%s=0x%x): enter: " fmt "\n", \
		_pth_trace_func, _pth_trace_tagname,	    \
		_pth_trace_tag, arg1, arg2), 0
#define TRACE_BEG3(lvl, name, tag, fmt, arg1, arg2, arg3)	    \
  _TRACE (lvl, name, tag);					    \
  _pth_debug (_pth_trace_level, "%s (%s=0x%x): enter: " fmt "\n", \
		_pth_trace_func, _pth_trace_tagname,	    \
		_pth_trace_tag, arg1, arg2, arg3), 0
#define TRACE_BEG4(lvl, name, tag, fmt, arg1, arg2, arg3, arg4)	    \
  _TRACE (lvl, name, tag);					    \
  _pth_debug (_pth_trace_level, "%s (%s=0x%x): enter: " fmt "\n", \
		_pth_trace_func, _pth_trace_tagname,	    \
		_pth_trace_tag, arg1, arg2, arg3, arg4), 0

#define TRACE(lvl, name, tag)						\
  _pth_debug (lvl, "%s (%s=0x%x): call\n",				\
		name, STRINGIFY (tag), (void *) tag), 0
#define TRACE0(lvl, name, tag, fmt)					\
  _pth_debug (lvl, "%s (%s=0x%x): call: " fmt "\n",			\
		name, STRINGIFY (tag), (void *) tag), 0
#define TRACE1(lvl, name, tag, fmt, arg1)			       \
  _pth_debug (lvl, "%s (%s=0x%x): call: " fmt "\n",		       \
		name, STRINGIFY (tag), (void *) tag, arg1), 0
#define TRACE2(lvl, name, tag, fmt, arg1, arg2)			       \
  _pth_debug (lvl, "%s (%s=0x%x): call: " fmt "\n",		       \
		name, STRINGIFY (tag), (void *) tag, arg1, arg2), 0
#define TRACE3(lvl, name, tag, fmt, arg1, arg2, arg3)		       \
  _pth_debug (lvl, "%s (%s=0x%x): call: " fmt "\n",		       \
		name, STRINGIFY (tag), (void *) tag, arg1, arg2,       \
		arg3), 0
#define TRACE6(lvl, name, tag, fmt, arg1, arg2, arg3, arg4, arg5, arg6)	\
  _pth_debug (lvl, "%s (%s=0x%x): call: " fmt "\n",			\
		name, STRINGIFY (tag), (void *) tag, arg1, arg2, arg3,	\
		arg4, arg5, arg6), 0

#define TRACE_ERR(err)							\
  err == 0 ? (TRACE_SUC ()) :						\
    (_pth_debug (_pth_trace_level, "%s (%s=0x%x): error: %s <%s>\n",	\
		   _pth_trace_func, _pth_trace_tagname,		\
		   _pth_trace_tag, pth_strerror (err),		\
		   pth_strsource (err)), (err))
/* The cast to void suppresses GCC warnings.  */
#define TRACE_SYSRES(res)						\
  res >= 0 ? ((void) (TRACE_SUC1 ("result=%i", res)), (res)) :		\
    (_pth_debug (_pth_trace_level, "%s (%s=0x%x): error: %s\n",	\
		   _pth_trace_func, _pth_trace_tagname,		\
		   _pth_trace_tag, strerror (errno)), (res))
#define TRACE_SYSERR(res)						\
  res == 0 ? ((void) (TRACE_SUC1 ("result=%i", res)), (res)) :		\
    (_pth_debug (_pth_trace_level, "%s (%s=0x%x): error: %s\n",	\
		   _pth_trace_func, _pth_trace_tagname,		\
		   _pth_trace_tag, strerror (res)), (res))

#define TRACE_SUC()						 \
  _pth_debug (_pth_trace_level, "%s (%s=0x%x): leave\n",       \
		_pth_trace_func, _pth_trace_tagname,	 \
		_pth_trace_tag), 0
#define TRACE_SUC0(fmt)							\
  _pth_debug (_pth_trace_level, "%s (%s=0x%x): leave: " fmt "\n",	\
		_pth_trace_func, _pth_trace_tagname,		\
		_pth_trace_tag), 0
#define TRACE_SUC1(fmt, arg1)						\
  _pth_debug (_pth_trace_level, "%s (%s=0x%x): leave: " fmt "\n",	\
		_pth_trace_func, _pth_trace_tagname,		\
		_pth_trace_tag, arg1), 0
#define TRACE_SUC2(fmt, arg1, arg2)					\
  _pth_debug (_pth_trace_level, "%s (%s=0x%x): leave: " fmt "\n",	\
		_pth_trace_func, _pth_trace_tagname,		\
		_pth_trace_tag, arg1, arg2), 0
#define TRACE_SUC5(fmt, arg1, arg2, arg3, arg4, arg5)			\
  _pth_debug (_pth_trace_level, "%s (%s=0x%x): leave: " fmt "\n",	\
		_pth_trace_func, _pth_trace_tagname,		\
		_pth_trace_tag, arg1, arg2, arg3, arg4, arg5), 0

#define TRACE_LOG(fmt)							\
  _pth_debug (_pth_trace_level, "%s (%s=0x%x): check: " fmt "\n",	\
		_pth_trace_func, _pth_trace_tagname,		\
		_pth_trace_tag), 0
#define TRACE_LOG1(fmt, arg1)						\
  _pth_debug (_pth_trace_level, "%s (%s=0x%x): check: " fmt "\n", \
		_pth_trace_func, _pth_trace_tagname,	    \
		_pth_trace_tag, arg1), 0
#define TRACE_LOG2(fmt, arg1, arg2)				    \
  _pth_debug (_pth_trace_level, "%s (%s=0x%x): check: " fmt "\n", \
		_pth_trace_func, _pth_trace_tagname,	    \
		_pth_trace_tag, arg1, arg2), 0
#define TRACE_LOG3(fmt, arg1, arg2, arg3)			    \
  _pth_debug (_pth_trace_level, "%s (%s=0x%x): check: " fmt "\n", \
		_pth_trace_func, _pth_trace_tagname,	    \
		_pth_trace_tag, arg1, arg2, arg3), 0
#define TRACE_LOG4(fmt, arg1, arg2, arg3, arg4)			    \
  _pth_debug (_pth_trace_level, "%s (%s=0x%x): check: " fmt "\n", \
		_pth_trace_func, _pth_trace_tagname,	    \
		_pth_trace_tag, arg1, arg2, arg3, arg4), 0
#define TRACE_LOG6(fmt, arg1, arg2, arg3, arg4, arg5, arg6)	    \
  _pth_debug (_pth_trace_level, "%s (%s=0x%x): check: " fmt "\n", \
		_pth_trace_func, _pth_trace_tagname,	    \
		_pth_trace_tag, arg1, arg2, arg3, arg4, arg5,	    \
		arg6), 0

#define TRACE_LOGBUF(buf, len)						\
  _pth_debug_buffer (_pth_trace_level, "%s (%s=0x%x): check: %s",	\
		       _pth_trace_func, _pth_trace_tagname,		\
		       _pth_trace_tag, buf, len)

#define TRACE_SEQ(hlp,fmt)						\
  _pth_debug_begin (&(hlp), _pth_trace_level,			\
		      "%s (%s=0x%x): check: " fmt,			\
		      _pth_trace_func, _pth_trace_tagname,		\
		      _pth_trace_tag)
#define TRACE_ADD0(hlp,fmt) \
  _pth_debug_add (&(hlp), fmt)
#define TRACE_ADD1(hlp,fmt,a) \
  _pth_debug_add (&(hlp), fmt, (a))
#define TRACE_ADD2(hlp,fmt,a,b) \
  _pth_debug_add (&(hlp), fmt, (a), (b))
#define TRACE_ADD3(hlp,fmt,a,b,c) \
  _pth_debug_add (&(hlp), fmt, (a), (b), (c))
#define TRACE_END(hlp,fmt) \
  _pth_debug_add (&(hlp), fmt); \
  _pth_debug_end (&(hlp))
#define TRACE_ENABLED(hlp) (!!(hlp))

#endif	/* DEBUG_H */


utils.h
/* utils.h - Internal definitions.
 * Copyright (C) 2010 g10 Code GmbH
 *
 * This file is part of W32PTH.
 *
 * W32PTH is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * W32PTH is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program; if not, see <http://www.gnu.org/licenses/>.
 */

#ifndef UTILS_H
#define UTILS_H

#ifdef HAVE_W32CE_SYSTEM
# include <gpg-error.h>  /* Required for gpg_err_set_errno().  */
#endif


static inline void
set_errno (int value)
{
#ifdef HAVE_W32CE_SYSTEM
  gpg_err_set_errno (value);
#else
  errno = value;
#endif
}

#define DIM(v)		     (sizeof(v)/sizeof((v)[0]))
#define DIMof(type,member)   DIM(((type *)0)->member)


/*-- w32-pth.c --*/
void *_pth_malloc (size_t n);
void *_pth_calloc (size_t n, size_t m);
void _pth_free (void *p);


#endif /*UTILS_H*/


w32-io.h
/* w32-io.h - interface to debugging functions
   Copyright (C) 2002, 2004, 2005, 2007 g10 Code GmbH
 
   This file is part of GPGME.

   GPGME is free software; you can redistribute it and/or modify it
   under the terms of the GNU Lesser General Public License as
   published by the Free Software Foundation; either version 2.1 of
   the License, or (at your option) any later version.
   
   GPGME is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.
   
   You should have received a copy of the GNU Lesser General Public
   License along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
   02111-1307, USA.  */

#ifndef W32_IO_H
#define W32_IO_H

/* Actually in w32-pth.c */
void *pth_malloc (size_t n);
void *pth_calloc (size_t n, size_t m);
void _pth_free (void *p);


/* w32-io.c */
void _pth_sema_subsystem_init (void);

/* For select.  */
HANDLE _pth_get_reader_ev (int fd);
HANDLE _pth_get_writer_ev (int fd);

int _pth_io_read (int fd, void *buffer, size_t count);
int _pth_io_write (int fd, const void *buffer, size_t count);


#endif	/* W32_IO_H */


autogen.sh
#! /bin/sh
# Run this to generate all the initial makefiles, etc. 
#
# Copyright (C) 2003 g10 Code GmbH
#
# This file is free software; as a special exception the author gives
# unlimited permission to copy and/or distribute it, with or without
# modifications, as long as this notice is preserved.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY, to the extent permitted by law; without even the
# implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

configure_ac="configure.ac"

cvtver () {
  awk 'NR==1 {split($NF,A,".");X=1000000*A[1]+1000*A[2]+A[3];print X;exit 0}'
}

check_version () {
    if [ `("$1" --version || echo "0") | cvtver` -ge "$2" ]; then
       return 0
    fi
    echo "**Error**: "\`$1\'" not installed or too old." >&2
    echo '           Version '$3' or newer is required.' >&2
    [ -n "$4" ] && echo '           Note that this is part of '\`$4\''.' >&2
    DIE="yes"
    return 1
}

DIE=no
FORCE=
if test x"$1" = x"--force"; then
  FORCE=" --force"
  shift
fi

# Begin list of optional variables sourced from ~/.gnupg-autogen.rc
w32_toolprefixes=
w32_extraoptions=
w32ce_toolprefixes=
w32ce_extraoptions=
amd64_toolprefixes=
# End list of optional variables sourced from ~/.gnupg-autogen.rc
# What follows are variables which are sourced but default to 
# environment variables or lacking them hardcoded values.
#w32root=
#w32ce_root=
#amd64root=

if [ -f "$HOME/.gnupg-autogen.rc" ]; then
    echo "sourcing extra definitions from $HOME/.gnupg-autogen.rc"
    . "$HOME/.gnupg-autogen.rc"
fi

# Convenience option to use certain configure options for some hosts.
myhost="" 
myhostsub=""
case "$1" in
    --build-w32)
        myhost="w32"
        ;;
    --build-w32ce)
        myhost="w32"
        myhostsub="ce"
        ;;
    --build*)
        echo "**Error**: invalid build option $1" >&2
        exit 1
        ;;
    *)
        ;;
esac



# ***** W32 build script *******
# Used to cross-compile for Windows.
if [ "$myhost" = "w32" ]; then
    tmp=`dirname $0`
    tsdir=`cd "$tmp"; pwd`
    shift
    if [ ! -f $tsdir/config.guess ]; then
        echo "$tsdir/config.guess not found" >&2
        exit 1
    fi
    build=`$tsdir/config.guess`

    case $myhostsub in
        ce)
          w32root="$w32ce_root"
          [ -z "$w32root" ] && w32root="$HOME/w32ce_root"
          toolprefixes="arm-mingw32ce"
          extra_options="--with-gpg-error-prefix=${w32root}"
          ;;
        *)
          [ -z "$w32root" ] && w32root="$HOME/w32root"
          toolprefixes="i686-w64-mingw32 i586-mingw32msvc i386-mingw32msvc"
          extra_options=""
          ;;
    esac
    echo "Using $w32root as standard install directory" >&2

    crossbindir=
    for host in $toolprefixes; do
        if ${host}-gcc --version >/dev/null 2>&1 ; then
            crossbindir=/usr/${host}/bin
            conf_CC="CC=${host}-gcc"
            break;
        fi
    done
    if [ -z "$crossbindir" ]; then
        echo "Cross compiler kit not installed" >&2
        if [ -z "$sub" ]; then 
          echo "Under Debian GNU/Linux, you may install it using" >&2
          echo "  apt-get install mingw32 mingw32-runtime mingw32-binutils" >&2 
        fi
        echo "Stop." >&2
        exit 1
    fi
   
   
    if [ -f "$tsdir/config.log" ]; then
        if ! head $tsdir/config.log | grep "$host" >/dev/null; then
            echo "Pease run a 'make distclean' first" >&2
            exit 1
        fi
    fi

    $tsdir/configure --enable-maintainer-mode  --prefix=${w32root}  \
            --host=${host} --build=${build} ${extra_options} "$@"

    exit $?
fi
# ***** end W32 build script *******



# Grep the required versions from configure.ac
autoconf_vers=`sed -n '/^AC_PREREQ(/ { 
s/^.*(\(.*\))/\1/p
q
}' ${configure_ac}`
autoconf_vers_num=`echo "$autoconf_vers" | cvtver`

automake_vers=`sed -n '/^min_automake_version=/ { 
s/^.*="\(.*\)"/\1/p
q
}' ${configure_ac}`
automake_vers_num=`echo "$automake_vers" | cvtver`

#gettext_vers=`sed -n '/^AM_GNU_GETTEXT_VERSION(/ { 
#s/^.*(\(.*\))/\1/p
#q
#}' ${configure_ac}`
#gettext_vers_num=`echo "$gettext_vers" | cvtver`


if [ -z "$autoconf_vers" -o -z "$automake_vers" ]
then
  echo "**Error**: version information not found in "\`${configure_ac}\'"." >&2
  exit 1
fi

# Allow to override the default tool names
AUTOCONF=${AUTOCONF_PREFIX}${AUTOCONF:-autoconf}${AUTOCONF_SUFFIX}
AUTOHEADER=${AUTOCONF_PREFIX}${AUTOHEADER:-autoheader}${AUTOCONF_SUFFIX}

AUTOMAKE=${AUTOMAKE_PREFIX}${AUTOMAKE:-automake}${AUTOMAKE_SUFFIX}
ACLOCAL=${AUTOMAKE_PREFIX}${ACLOCAL:-aclocal}${AUTOMAKE_SUFFIX}

#GETTEXT=${GETTEXT_PREFIX}${GETTEXT:-gettext}${GETTEXT_SUFFIX}
#MSGMERGE=${GETTEXT_PREFIX}${MSGMERGE:-msgmerge}${GETTEXT_SUFFIX}

if check_version $AUTOCONF $autoconf_vers_num $autoconf_vers ; then
    check_version $AUTOHEADER $autoconf_vers_num $autoconf_vers autoconf
fi
if check_version $AUTOMAKE $automake_vers_num $automake_vers; then
  check_version $ACLOCAL $automake_vers_num $autoconf_vers automake
fi
#if check_version $GETTEXT $gettext_vers_num $gettext_vers; then
#  check_version $MSGMERGE $gettext_vers_num $gettext_vers gettext
#fi

if test "$DIE" = "yes"; then
    cat <<EOF

Note that you may use alternative versions of the tools by setting 
the corresponding environment variables; see README.SVN for details.
                   
EOF
    exit 1
fi

echo "Running aclocal -I m4 ${ACLOCAL_FLAGS:+$ACLOCAL_FLAGS }..."
$ACLOCAL -I m4  $ACLOCAL_FLAGS
echo "Running autoheader..."
$AUTOHEADER
echo "Running automake --gnu ..."
$AUTOMAKE --gnu;
echo "Running autoconf${FORCE} ..."
$AUTOCONF${FORCE}

echo "You may now run 
  ./autogen.sh --build-w32 && make
"


pth.h
/* pth.h - GNU Pth emulation for W32 (MS Windows).
 * Copyright (c) 1999-2003 Ralf S. Engelschall <rse@engelschall.com>
 * Copyright (C) 2004, 2006, 2007, 2008, 2010 g10 Code GmbH
 *
 * This file is part of W32PTH.
 *
 * W32PTH is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * W32PTh is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program; if not, see <http://www.gnu.org/licenses/>.
 *
 * ------------------------------------------------------------------
 * This code is based on Ralf Engelschall's GNU Pth, a non-preemptive
 * thread scheduling library which can be found at
 * http://www.gnu.org/software/pth/.
 */

#ifndef _W32_PTH_H
#define _W32_PTH_H

#ifdef _PTH_H_
#error w32pth conflict.  A vanilla pth.h has already been included.
#endif

#include <winsock2.h> /* Newer mingw version require this to be
                         included before windows.h.  */

/* #include <windows.h>  /\* We need this for sockaddr et al.  FIXME: too */
/*                          heavyweight - may be we should factor such */
/*                          code out to a second header and adjust all */
/*                          user files to include it only if required. *\/ */

/* Mingw64 defines sigset_t only of _POSIX is defined.  We don't want
   to do define this macro because we are a public header.  Thus we
   use a bit of mingw64 internal knowledge to declare sigset_t.  */
#include <sys/types.h>
#if __MINGW64_VERSION_MAJOR >= 2 && !defined(_POSIX)
typedef _sigset_t	sigset_t;
#endif /* mingw64 >= 2 */


#ifndef W32_PTH_HANDLE_INTERNAL
#define W32_PTH_HANDLE_INTERNAL  int
#endif

/* These are needed for pipe support.  Sigh.  */
int pth_pipe (int filedes[2], int inherit_idx);
int pth_close (int fd);


/* We need to define value for the how argument of pth_sigmask.  This
   is required because Mingw does not yet define sigprocmask.  We use
   an enum to error out if Mingw eventually defines them.  Also define
   the sigset_t. */
#ifdef __MINGW32__
enum 
  {
    SIG_BLOCK = 0,
    SIG_UNBLOCK = 1,
    SIG_SETMASK = 2
  };
#endif /*__MINGW32__*/

/* Filedescriptor blocking modes.  */
enum
  {
    PTH_FDMODE_ERROR = -1,
    PTH_FDMODE_POLL  =  0,
    PTH_FDMODE_BLOCK,
    PTH_FDMODE_NONBLOCK
  };


/* Mutex values. */
#define PTH_MUTEX_INITIALIZED  (1<<0)
#define PTH_MUTEX_LOCKED       (1<<1)

/* Note: We can't do static initialization, thus we don't define the
   initializer PTH_MUTEX_INIT.  */

/* Read-write lock values.  */
enum
  {
    PTH_RWLOCK_RD,
    PTH_RWLOCK_RW
  };

/* Note: We can't do static initialization, thus we don't define the
   initializer PTH_RWLOCK_INIT.  */


#define PTH_KEY_INIT	       (1<<0)


/* Event subject classes. */
#define PTH_EVENT_FD           (1<<1)
#define PTH_EVENT_SELECT       (1<<2)
#define PTH_EVENT_SIGS         (1<<3)
#define PTH_EVENT_TIME         (1<<4)
#define PTH_EVENT_MSG          (1<<5)
#define PTH_EVENT_MUTEX        (1<<6)
#define PTH_EVENT_COND         (1<<7)
#define PTH_EVENT_TID          (1<<8)
#define PTH_EVENT_FUNC         (1<<9)
#define PTH_EVENT_HANDLE       (1<<10)  /* A generic waitable W32 HANDLE. */

/* Event occurrence restrictions. */
#define PTH_UNTIL_OCCURRED     (1<<11)
#define PTH_UNTIL_FD_READABLE  (1<<12)
#define PTH_UNTIL_FD_WRITEABLE (1<<13)
#define PTH_UNTIL_FD_EXCEPTION (1<<14)
#define PTH_UNTIL_TID_NEW      (1<<15)
#define PTH_UNTIL_TID_READY    (1<<16)
#define PTH_UNTIL_TID_WAITING  (1<<17)
#define PTH_UNTIL_TID_DEAD     (1<<18)


/* Event structure handling modes. */
#define PTH_MODE_REUSE         (1<<20)
#define PTH_MODE_CHAIN         (1<<21)
#define PTH_MODE_STATIC        (1<<22)


/* Attribute commands for pth_attr_get and pth_attr_set(). */
enum
  {
    PTH_ATTR_PRIO,           /* RW [int]           Priority of thread.  */
    PTH_ATTR_NAME,           /* RW [char *]        Name of thread.  */
    PTH_ATTR_JOINABLE,       /* RW [int]           Thread detachment type.  */
    PTH_ATTR_CANCEL_STATE,   /* RW [unsigned int]  Thread cancellation state.*/
    PTH_ATTR_STACK_SIZE,     /* RW [unsigned int]  Stack size. */
    PTH_ATTR_STACK_ADDR,     /* RW [char *]        Stack lower address. */
    PTH_ATTR_DISPATCHES,     /* RO [int]           Total number of
                                                   thread dispatches. */
    PTH_ATTR_TIME_SPAWN,     /* RO [pth_time_t]    Time thread was spawned.  */
    PTH_ATTR_TIME_LAST,      /* RO [pth_time_t]    Time thread was
                                                   last dispatched.  */
    PTH_ATTR_TIME_RAN,       /* RO [pth_time_t]    Time thread was running.  */
    PTH_ATTR_START_FUNC,     /* RO [void *(*)(void *)] Thread start function.*/
    PTH_ATTR_START_ARG,      /* RO [void *]        Thread start argument.  */
    PTH_ATTR_STATE,          /* RO [pth_state_t]   Scheduling state. */
    PTH_ATTR_EVENTS,         /* RO [pth_event_t]   Events the thread 
                                                   is waiting for.  */
    PTH_ATTR_BOUND           /* RO [int]           Whether object is 
                                                   bound to thread. */
  };



/* Queries for pth_ctrl(). */
#define PTH_CTRL_GETAVLOAD            (1<<1)
#define PTH_CTRL_GETPRIO              (1<<2)
#define PTH_CTRL_GETNAME              (1<<3)
#define PTH_CTRL_GETTHREADS_NEW       (1<<4)
#define PTH_CTRL_GETTHREADS_READY     (1<<5)
#define PTH_CTRL_GETTHREADS_RUNNING   (1<<6)
#define PTH_CTRL_GETTHREADS_WAITING   (1<<7)
#define PTH_CTRL_GETTHREADS_SUSPENDED (1<<8)
#define PTH_CTRL_GETTHREADS_DEAD      (1<<9)
#define PTH_CTRL_DUMPSTATE            (1<<10)

#define PTH_CTRL_GETTHREADS           (  PTH_CTRL_GETTHREADS_NEW       \
                                       | PTH_CTRL_GETTHREADS_READY     \
                                       | PTH_CTRL_GETTHREADS_RUNNING   \
                                       | PTH_CTRL_GETTHREADS_WAITING   \
                                       | PTH_CTRL_GETTHREADS_SUSPENDED \
                                       | PTH_CTRL_GETTHREADS_DEAD        )


/* Event status codes. */
typedef enum
  {
    PTH_STATUS_PENDING,
    PTH_STATUS_OCCURRED,
    PTH_STATUS_FAILED
  }
pth_status_t;


/* Event deallocation types. */
enum 
  {
    PTH_FREE_THIS,
    PTH_FREE_ALL 
  };


/* The Pth thread handle object.  */
typedef void *pth_t;


/* The Mutex object.  */
typedef W32_PTH_HANDLE_INTERNAL pth_mutex_t;

/* The read-write lock object.  */
typedef W32_PTH_HANDLE_INTERNAL pth_rwlock_t;


/* The Event object.  */
struct pth_event_s;
typedef struct pth_event_s *pth_event_t;


/* The Attribute object.  */
struct pth_attr_s;
typedef struct pth_attr_s *pth_attr_t;


/* The Key object.  */
typedef int pth_key_t;


/* The Pth time object.  */
typedef struct timeval pth_time_t;


/* Function prototypes. */
int pth_init (void);
int pth_kill (void);
long pth_ctrl (unsigned long query, ...);

int pth_read_ev (int fd, void *buffer, size_t size, pth_event_t ev);
int pth_read (int fd,  void *buffer, size_t size);
int pth_write_ev (int fd, const void *buffer, size_t size, pth_event_t ev);
int pth_write (int fd, const void *buffer, size_t size);

int pth_select (int nfds, fd_set *rfds, fd_set *wfds, fd_set *efds,
		const struct timeval *timeout);
int pth_select_ev (int nfds, fd_set *rfds, fd_set *wfds, fd_set *efds,
                   const struct timeval * timeout, pth_event_t ev_extra);

int pth_accept (int fd, struct sockaddr *addr, int *addrlen);
int pth_accept_ev (int fd, struct sockaddr *addr, int *addrlen,
                   pth_event_t hd);

int pth_connect (int fd, struct sockaddr *name, int namelen);


int pth_mutex_release (pth_mutex_t *hd);
int pth_mutex_acquire(pth_mutex_t *hd, int try_only, pth_event_t ev_extra);
int pth_mutex_init (pth_mutex_t *hd);
/* We need this under windows, otherwise we would leak handles.  */
int pth_mutex_destroy (pth_mutex_t *hd);

int pth_rwlock_init (pth_rwlock_t *rwlock);
int pth_rwlock_acquire (pth_rwlock_t *rwlock, int op, int try, pth_event_t ev);
int pth_rwlock_release (pth_rwlock_t *rwlock);


pth_attr_t pth_attr_new (void);
int pth_attr_destroy (pth_attr_t hd);
int pth_attr_set (pth_attr_t hd, int field, ...);

pth_t pth_spawn (pth_attr_t hd, void *(*func)(void *), void *arg);
pth_t pth_self (void);
int pth_join (pth_t hd, void **value);
int pth_abort (pth_t hd);
void pth_exit (void *value);

unsigned int pth_waitpid (unsigned int, int *status, int options);
int pth_wait (pth_event_t hd);

int pth_sleep (int n);
int pth_usleep (unsigned int usec);
pth_time_t pth_timeout (long sec, long usec);



pth_event_t pth_event_isolate (pth_event_t hd);
int pth_event_free (pth_event_t hd, int mode);
int pth_event_status (pth_event_t hd);
int pth_event_occurred (pth_event_t hd);
pth_event_t pth_event_concat (pth_event_t ev, ...);
pth_event_t pth_event (unsigned long spec, ...);


int pth_sigmask (int how, const sigset_t *set, sigset_t *old);

int pth_yield (pth_t tid);

void pth_enter (void);
void pth_leave (void);


/* Special W32 function to cope with the problem that pth_self returns
   just a pseudo handle which is not very useful for debugging.  */
unsigned long pth_thread_id (void);
#define PTH_HAVE_PTH_THREAD_ID 1


/*-- pth_util.c --*/

/* void sigemptyset (struct sigset_s * ss); */

/* int sigaddset (struct sigset_s * ss, int signo); */



#endif /*_W32_PTH_H*/
